sales/src/execute.rs

```diff
@@ -0,0 +1,209 @@
+use std::{cell::RefCell, rc::Rc};
+
+use burnt_glue::response::Response;
+use cosmwasm_std::{
+    BankMsg, Coin, CosmosMsg, CustomMsg, Deps, DepsMut, Env, MessageInfo, Timestamp, Uint64,
+};
+use cw721_base::{state::TokenInfo, MintMsg};
+use cw_storage_plus::Item;
+use ownable::Ownable;
+use sellable::Sellable;
+use serde::{de::DeserializeOwned, Serialize};
+
+use crate::{errors::ContractError, state::PrimarySale, Sales};
+
+impl<'a, T, C, E, Q> Sales<'a, T, C, E, Q>
+where
+    T: Serialize + DeserializeOwned + Clone,
+    Q: CustomMsg,
+    E: CustomMsg,
+    C: CustomMsg,
+{
+    pub fn new(
+        sellable_module: Rc<RefCell<Sellable<'a, T, C, E, Q>>>,
+        primary_sales: Item<'a, Vec<PrimarySale>>,
+    ) -> Self {
+        Self {
+            sellable: sellable_module,
+            primary_sales,
+        }
+    }
+
+    pub fn add_primary_sales(
+        &mut self,
+        total_supply: Uint64,
+        start_time: Uint64,
+        end_time: Uint64,
+        price: Vec<Coin>,
+        deps: &mut DepsMut,
+        env: Env,
+        info: &MessageInfo,
+    ) -> Result<Response, ContractError> {
+        let ownable = &self.sellable.borrow().ownable;
+        check_ownable(&deps.as_ref(), &env, &info, &ownable.borrow())?;
+        // make sure no active primary sale
+        let mut primary_sales = self.primary_sales.load(deps.storage).unwrap();
+        for sale in &primary_sales {
```

Comment:
We should reuse the query function `active_primary_sales`. Since the type signature of that is a `StdResult`, you may have to use `map_err`

Reply:
Also we don't mind if there is a current primary sale going on, so long as there is _no conflict_ between the dates of the sale being added and _any existing sales_

Reply:
Oh really! I thought we could only add one active item at a time 

Does state serialization keep the order of items in a Vec? 
if it does not, then I have to loop through the whole array to find the earliest primary sale then

Reply:
For now, I assume they preserve order and assume our Vec of primary sales are ordered since we push into the array in an ordered manner.